File: 01-Programming-Languages/R/Data-Structures/Graph/graph_adjacency_list.R
# Problem: Graph as Adjacency List
# Description: Build and traverse a graph using adjacency list representation.
#
# Approach:
# - Store neighbors for each vertex in a list
# - Provide add_edge and display methods
#
# Time Complexity: O(V + E) for traversal
# Space Complexity: O(V + E)
#
# Example:
# Input: add_edge(1,2), add_edge(1,3), add_edge(2,4)
# Output: Graph with 4 vertices

Graph <- function(vertices) {
  list(vertices = vertices, adj = vector("list", vertices))
}

add_edge <- function(graph, u, v) {
  graph$adj[[u]] <- c(graph$adj[[u]], v)
  graph$adj[[v]] <- c(graph$adj[[v]], u)
  graph
}

display_graph <- function(graph) {
  for (i in 1:graph$vertices) {
    cat(i, "->", graph$adj[[i]], "\n")
  }
}

g <- Graph(5)
g <- add_edge(g, 1, 2)
g <- add_edge(g, 1, 3)
g <- add_edge(g, 2, 4)
g <- add_edge(g, 3, 5)
g <- add_edge(g, 4, 5)

cat("Adjacency List:\n")
display_graph(g)

File: 01-Programming-Languages/Scala/Data-Structures/Graph/GraphAdjacencyList.scala
// Problem: Graph as Adjacency List
// Description: Build and traverse a graph using adjacency list representation.
//
// Approach:
// - Store neighbors for each vertex in an array of lists
// - Provide addEdge and display methods
//
// Time Complexity: O(V + E) for traversal
// Space Complexity: O(V + E)
//
// Example:
// Input: addEdge(1,2), addEdge(1,3), addEdge(2,4)
// Output: Graph with 4 vertices

object GraphAdjacencyList {
  class Graph(vertices: Int) {
    val adj: Array[scala.collection.mutable.ListBuffer[Int]] =
      Array.fill(vertices)(scala.collection.mutable.ListBuffer[Int]())

    def addEdge(u: Int, v: Int): Unit = {
      adj(u) += v
      adj(v) += u
    }

    def display(): Unit = {
      for (i <- 0 until vertices) {
        println(s"$i -> ${adj(i).mkString(", ")}")
      }
    }
  }

  def main(args: Array[String]): Unit = {
    val g = new Graph(5)
    g.addEdge(0, 1)
    g.addEdge(0, 2)
    g.addEdge(1, 3)
    g.addEdge(2, 4)
    g.addEdge(3, 4)
    println("Adjacency List:")
    g.display()
  }
}
